src/cmap/conn/stream.rs

```diff
@@ -38,27 +38,53 @@ pub(super) enum Stream {
     ),
 }
 
+fn try_connect(address: &SocketAddr, timeout: Duration) -> Result<TcpStream> {
+    // The URI options spec requires that the default connect timeout is 10 seconds, but that 0
+    // should indicate no timeout.
+    let stream = if timeout == Duration::from_secs(0) {
+        TcpStream::connect(address)?
+    } else {
+        TcpStream::connect_timeout(address, timeout)?
+    };
+
+    Ok(stream)
+}
+
+fn connect_stream(address: &StreamAddress, connect_timeout: Option<Duration>) -> Result<TcpStream> {
+    let timeout = connect_timeout.unwrap_or(DEFAULT_CONNECT_TIMEOUT);
+
+    let mut socket_addrs: Vec<_> = address.to_socket_addrs()?.collect();
+
+    if socket_addrs.is_empty() {
+        return Err(ErrorKind::NoDnsResults(address.clone()).into());
+    }
+
+    socket_addrs.sort_by_key(|addr| if addr.is_ipv4() { 0 } else { 1 });
```

Comment:
why was this methodology chosen over of happy eyeballs? It might be worth recording the rationale in `connect_stream`'s docstring.

Reply:
We discussed this in a triage meeting a few weeks ago (which I think you were present for, but I could be wrong). Matt seemed unconvinced that the overheard of happy eyeballs (spawning additional threads) and the complexity of implementation were worth the potential gains from not checking them in sequence. He also brought up that C is the only driver that uses happy eyeballs currently, if I recall correctly.

Reply:
Sounds good--could we leave a docstring or comment indicating that?

Reply:
This line is existing code from a previous commit; the only change to it is that it was moved to a different function. Is there a reason why it sticks out to you now when it didn't before?

Reply:
We put it in before as a temporary measure until we fully implemented happy eyeballs. Now that we've decided not to implement happy eyeballs and commit to the temporary measure, it makes sense to me for us to record why.

Reply:
I think we discussed this one offline, and decided to record that "other approaches were entertained, but ultimately we decided to do what other drivers are doing"